### PR TITLE
NUP-2458 fix classifier serialization deque index out of bounds

### LINKED ISSUE #3686: Fix SDRClassifier capnp serialization bug: IndexError: deque index out of range
```
======================================================================
ERROR: testSimplePYModelSerializationNoValidation (__main__.HTMPredictionModelSerializationTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "htmpredictionmodel_serialization_test.py", line 262, in testSimplePYModelSerializationNoValidation
    self._runSimpleModelSerializationDeserialization(PY_MODEL_PARAMS)
  File "htmpredictionmodel_serialization_test.py", line 238, in _runSimpleModelSerializationDeserialization
    srcModel.write(builderProto)
  File "/Users/vkruglikov/Library/Python/2.7/lib/python/site-packages/nupic/frameworks/opf/htm_prediction_model.py", line 1320, in write
    self._netInfo.net.write(proto.network)
  File "/Users/vkruglikov/Library/Python/2.7/lib/python/site-packages/nupic/bindings/engine_internal.py", line 1294, in write
    reader = NetworkProto.from_bytes(self._writeAsCapnpPyBytes()) # copy
  File "/Users/vkruglikov/Library/Python/2.7/lib/python/site-packages/nupic/bindings/engine_internal.py", line 1310, in _writeAsCapnpPyBytes
    return _engine_internal.Network__writeAsCapnpPyBytes(self)
  File "/Users/vkruglikov/Library/Python/2.7/lib/python/site-packages/nupic/bindings/engine_internal.py", line 2974, in writePyRegion
    getattr(region, methodName)(builderProto)
  File "/Users/vkruglikov/Library/Python/2.7/lib/python/site-packages/nupic/bindings/regions/PyRegion.py", line 346, in write
    self.writeToProto(regionImpl)
  File "/Users/vkruglikov/Library/Python/2.7/lib/python/site-packages/nupic/regions/sdr_classifier_region.py", line 331, in writeToProto
    self._sdrClassifier.write(proto.sdrClassifier)
  File "/Users/vkruglikov/Library/Python/2.7/lib/python/site-packages/nupic/algorithms/sdr_classifier.py", line 429, in write
    subPatternProto = patternProto.init(i, len(self._patternNZHistory[i][1]))
IndexError: deque index out of range
```

### PR DESCRIPTION
**Pull request by vitaly-krugl**
Fixes #3686

NOTE: the `htmpredictionmodel_serialization_test.py` test in this PR is very rudimentary, just enough to demonstrate the failure. The test will evolve in subsequent pull requests related to testing of the serialization of `HTMPredictionModel`.

### CODE REVIEW

File: tests/unit/nupic/algorithms/sdr_classifier_test.py [hunk 1]
```diff
@@ -841,11 +840,25 @@ def testWriteRead(self):
...
       for i in xrange(len(c1._actualValues)):
         self.assertAlmostEqual(result1[key][i], result2[key][i], 5)
 
 
+  @unittest.skipUnless(
+      capnp, "pycapnp is not installed, skipping serialization test.")
+  def testWriteRead(self):
+    self._doWriteReadChecks(computeBeforeSerializing=True)
+
+  @unittest.skip("Good test to have, but fails on i out of bounds in "
```
Comment (vitaly-krugl) [inline tests/unit/nupic/algorithms/sdr_classifier_test.py]: This test reproduces the "deque index out of bounds" in write before this PR's fix. However, after the fix it fails in a different, likely unrelated, way that needs to be debugged, so skipping for now:
```
======================================================================
ERROR: testWriteReadNoComputeBeforeSerializing (__main__.SDRClassifierTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "sdr_classifier_test.py", line 859, in testWriteReadNoComputeBeforeSerializing
    self._doWriteReadChecks(computeBeforeSerializing=False)
  File "sdr_classifier_test.py", line 846, in _doWriteReadChecks
    self.assertAlmostEqual(result1[key][i], result2[key][i], 5)
IndexError: list index out of range
```

**Comment by vitaly-krugl** [comment]
@scottpurdy, please review

File: tests/integration/nupic/opf/htmpredictionmodel_serialization_test.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+        'anomalyParams': {  u'anomalyCacheRecords': None,
+                            u'autoDetectThreshold': None,
+    u'autoDetectWaitRecords': 2184},
+
+        'trainSPNetOnlyIfRequested': False,
+        },
+}
+
+
+class HTMPredictionModelSerializationTest(unittest.TestCase):
```
Comment (vitaly-krugl) [inline tests/integration/nupic/opf/htmpredictionmodel_serialization_test.py]: NOTE: This test is very rudimentary, just enough to demonstrate the failure being addressed by this pull request. The test will evolve in subsequent pull requests related to testing of the serialization of `HTMPredictionModel`.

File: tests/unit/nupic/algorithms/sdr_classifier_test.py [hunk 3]
```diff
@@ -841,11 +840,25 @@ def testWriteRead(self):
...
+                 "result1[key][i]. Need to debug.")
```
Comment (vitaly-krugl) [inline tests/unit/nupic/algorithms/sdr_classifier_test.py]: This test reproduces the "deque index out of bounds" in write before this PR's fix. However, with the fix it fails in a different, likely unrelated, way that needs to be debugged, so skipping for now:
```
======================================================================
ERROR: testWriteReadNoComputeBeforeSerializing (__main__.SDRClassifierTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "sdr_classifier_test.py", line 859, in testWriteReadNoComputeBeforeSerializing
    self._doWriteReadChecks(computeBeforeSerializing=False)
  File "sdr_classifier_test.py", line 846, in _doWriteReadChecks
    self.assertAlmostEqual(result1[key][i], result2[key][i], 5)
IndexError: list index out of range
```

File: tests/integration/nupic/opf/htmpredictionmodel_serialization_test.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+"""
+This module tests capnp serialization of HTMPredictionModel.
+"""
+
+import datetime
+import numpy.random
```
Comment (scottpurdy) [inline tests/integration/nupic/opf/htmpredictionmodel_serialization_test.py]: Can you use `nupic.bindings.math.Random` instead?

Comment (vitaly-krugl) [inline tests/integration/nupic/opf/htmpredictionmodel_serialization_test.py]: @scottpurdy: The test code is using `numpy.random.uniform`, apparently not available in `nupic.bindings.math.Random`.

**Pull request merged**